db/crud.go

```diff
@@ -489,8 +489,21 @@ func (db *DatabaseContext) getRevision(doc *document, revid string) (Body, error
 	body[BodyId] = doc.ID
 	body[BodyRev] = revid
 
-	if doc.Attachments != nil {
-		body[BodyAttachments] = doc.Attachments
+	attachments := make(map[string]interface{})
+
+	revGeneration, _ := base.ToInt64(genOfRevID(revid))
+	for name, attachment := range doc.Attachments {
+		attachmentMeta, ok := attachment.(map[string]interface{})
+		if ok {
+			attachmentRevpos, ok := base.ToInt64(attachmentMeta["revpos"])
+			if ok && attachmentRevpos <= revGeneration {
+				attachments[name] = attachment
+			}
+		}
+	}
+
+	if len(attachments) > 0 {
+		body[BodyAttachments] = attachments
```

Comment:
This turns out to be a more subtle bug than I'd previously realized, but after some testing with your new unit test locally I think I've got a clearer understanding of what's changed since 2.1.

You're correct in identifying that lines 492/493 are a problem in the current master - they stamp the attachments associated with the *active* revision into the body of whichever revision is being requested, which isn't necessarily the active revision.  However, I don't think the fix you've got is necessary/sufficient - by basing its calculations on generation only, it still has the potential to include incorrect attachments from conflicting rev branches.

In fact, if we are retrieving anything but the current revision's body, the _attachments should already be stamped into the body - they get included before we write OldRevisionJSON, and also in the case where we're storing conflicting revision bodies in the document metadata.  Given that, I think the fix is just to avoid overwriting with the current attachments unless we're asking for the current rev, by changing the old line 492 to:
``
if doc.CurrentRev == revid && doc.Attachments != nil
``